src/pass/detect_device.cc

```diff
@@ -0,0 +1,30 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file detect_device.cc
+ */
+
+#include <tvm/ir_pass.h>
+#include <tvm/ir_mutator.h>
+#include "../pass/ir_util.h"
+
+namespace tvm {
+namespace ir {
+
+class DetectDevice : public IRMutator {
+ public:
```

Comment:
IRMutator is not necessary, if we want to support this one, we can just inline the logic to MarkDevice